src/meta/src/stream/scale.rs

```diff
@@ -0,0 +1,827 @@
+                    }
+                }
+                _ => unimplemented!(),
+            }
+        }
+
+        let mut upstream_actors_to_remove = HashSet::new();
+        for upstream_fragment_id in &upstream_fragment_ids {
+            if let Some(actors_to_remove) = fragment_actors_to_remove.get(upstream_fragment_id) {
+                upstream_actors_to_remove.extend(actors_to_remove.keys().cloned());
+            }
+        }
+
+        // update upstream actor ids
+        new_actor
+            .upstream_actor_id
+            .drain_filter(|actor_id| upstream_actors_to_remove.contains(actor_id));
+
+        for upstream_fragment_id in &upstream_fragment_ids {
+            if let Some(upstream_actors_to_create) =
+                fragment_actors_to_create.get(upstream_fragment_id)
+            {
+                new_actor
+                    .upstream_actor_id
+                    .extend(upstream_actors_to_create.keys().cloned())
+            }
+        }
+
+        fn replace_merge_node_upstream(
+            stream_node: &mut StreamNode,
+            fragment_actors_to_remove: &HashMap<FragmentId, BTreeMap<ActorId, ParallelUnitId>>,
+            fragment_actors_to_create: &HashMap<FragmentId, BTreeMap<ActorId, ParallelUnitId>>,
+        ) {
+            if let Some(NodeBody::Merge(s)) = stream_node.node_body.as_mut() {
+                if let Some(upstream_actors_to_remove) =
+                    fragment_actors_to_remove.get(&s.upstream_fragment_id)
+                {
+                    s.upstream_actor_id
+                        .drain_filter(|actor_id| upstream_actors_to_remove.contains_key(actor_id));
+                }
+
+                if let Some(upstream_actors_to_create) =
+                    fragment_actors_to_create.get(&s.upstream_fragment_id)
+                {
+                    s.upstream_actor_id
+                        .extend(upstream_actors_to_create.keys().cloned());
+                }
+            }
+
+            for child in &mut stream_node.input {
+                replace_merge_node_upstream(
+                    child,
+                    fragment_actors_to_remove,
+                    fragment_actors_to_create,
+                );
+            }
+        }
+
+        if let Some(node) = new_actor.nodes.as_mut() {
+            replace_merge_node_upstream(node, fragment_actors_to_remove, fragment_actors_to_create);
+        }
+
+        // update downstream actor ids
+        for dispatcher in &mut new_actor.dispatcher {
+            let downstream_fragment_ids: HashSet<_> = dispatcher
+                .downstream_actor_id
+                .iter()
+                .filter_map(|actor_id| actor_map.get(actor_id).map(|actor| actor.fragment_id))
+                .collect();
+            assert_eq!(downstream_fragment_ids.len(), 1);
+            let downstream_fragment_id = downstream_fragment_ids.into_iter().next().unwrap();
```

Comment:
```suggestion
            let downstream_fragment_id = ..filter_map().dedup().exactly_one().unwrap();
```

We may also directly maintain the downstream fragment id in the future. 🤔